8248238: Implementation: JEP 388: Windows AArch64 Support

rnkovacs opened the pull request:
Main commit of the Windows/AArch64 port.

Also fixes [JDK-8272181](https://bugs.openjdk.java.net/browse/JDK-8272181) by adding [`src/hotspot/os_cpu/windows_aarch64/pauth_windows_aarch64.inline.hpp`](https://github.com/openjdk/jdk11u-dev/pull/301/files#diff-235302dc4618158c4f75771ebbb6f0e4beeee5daf0a537b03ee4253a711268c6), which is needed for a successful Windows-AArch64 build.

Depends on #274 and #299.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issues
 * [JDK-8248238](https://bugs.openjdk.java.net/browse/JDK-8248238): Implementation: JEP 388: Windows AArch64 Support
 * [JDK-8272181](https://bugs.openjdk.java.net/browse/JDK-8272181): Windows-AArch64:Backport fix of `Backtracing broken on PAC enabled systems`


### Reviewers
 * [Bernhard Urban-Forster](https://openjdk.java.net/census#burban) (@lewurm - no project role) ⚠️ Review applies to 7ae8001a1cb4be2c78d58fbe929dc22afd69e417
 * [Andrew Haley](https://openjdk.java.net/census#aph) (@theRealAph - **Reviewer**) ⚠️ Review applies to 697c339853a57fad73539fe15fc0e12d242177a5
 * [Andrew Dinn](https://openjdk.java.net/census#adinn) (@adinn - **Reviewer**) ⚠️ Review applies to 697c339853a57fad73539fe15fc0e12d242177a5
 * [Monica Beckwith](https://openjdk.java.net/census#mbeckwit) (@mo-beck - no project role) ⚠️ Review applies to d33ac7612cc59feb298b0183fed938439fdce7b5


### Contributors
 * Monica Beckwith `<mbeckwit@openjdk.org>`
 * Ludovic Henry `<luhenry@openjdk.org>`
 * Bernhard Urban-Forster `<burban@openjdk.org>`

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk11u-dev pull/301/head:pull/301` \
`$ git checkout pull/301`

Update a local copy of the PR: \
`$ git checkout pull/301` \
`$ git pull https://git.openjdk.java.net/jdk11u-dev pull/301/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 301`

View PR using the GUI difftool: \
`$ git pr show -t 301`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk11u-dev/pull/301.diff">https://git.openjdk.java.net/jdk11u-dev/pull/301.diff</a>

</details>

rnkovacs commented:
/contributor add Monica Beckwith <mbeckwit@openjdk.org>
/contributor add Ludovic Henry <luhenry@openjdk.org>
/contributor add Bernhard Urban-Forster <burban@openjdk.org>

rnkovacs commented:
/issue add 8272181

RealCLanger commented:
Hi @rnkovacs, can you split these into separate, dependent Pull requests?
E.g. #274 as first in the row. Then, open another PR in favor of #299 that has branch pr/274 as target. And then the 3 commits included in this one also as separate PRs, targeting the appropriate predecessor PR?

rnkovacs commented:
Hi @RealCLanger, I attempted to do that but apparently failed - I forgot to change the target branch. Do I need to close #299 and this one and open completely new PRs?

rnkovacs commented:
Seems like I could do the change on the GUI. Could you please check if it's correct now?

RealCLanger commented:
> Seems like I could do the change on the GUI. Could you please check if it's correct now?

Yes, I think this looks good now, thanks.

VladimirKempik commented:
One question, about R18 exclusion part.
In the light of recent issues with r18 being finally used by OS for its own needs ( since macos12b6), I have made some tests.  ( more about issues here https://youtrack.jetbrains.com/issue/JBR-3715)
I took part of this PR which belongs to R18 usage elimination and applied it to zulu11 ( reverting older version of patch first which wasn't work with C2 compiled code)
This fixed zulu11 under macos12b6 and I was happy about that.
However I have noticed one strange thing
I did this:
bin/java -XX:+UnlockDiagnosticVMOptions -XX:+PrintAssembly -jar demo/jfc/J2Ddemo/J2Ddemo.jar > output.log
then this:
cat output.log | grep x18 | grep -v 0x18
and found some usage of x18:
0x000000010bcafc80: stp	x18, x19, [sp, #144]
 0x000000010bcb0b60: stp	x18, x19, [sp, #144]
 0x000000010bcb1590: stp	x18, x19, [sp, #144]
 0x000000010bcb2470: stp	x18, x19, [sp, #144]
 0x000000010bcb3dd0: stp	x18, x19, [sp, #144]
 0x000000010bcb4cd0: stp	x18, x19, [sp, #144]
 0x000000010bcb5ae0: stp	x18, x19, [sp, #144]

it's just reading of x18, not writing, however it might indicate something is still missing there.
this doesn't happen on jdk17ea (on 17ea there is a different strange thing with x18)

Could you try to replicate this result with win-arm64  build  of jdk11?

VladimirKempik commented:
An example, it's always in Stub Code
`[Exception Handler]
[Stub Code]
 0x0000000109fe4340: adrp   x8, 0x0000000109b39000
                        ;  {no_reloc}
 0x0000000109fe4344: add    x8, x8, #0x580
 0x0000000109fe4348: blr    x8
 0x0000000109fe434c: stp    x0, x1, [sp, #-256]!
 0x0000000109fe4350: stp    x2, x3, [sp, #16]
 0x0000000109fe4354: stp    x4, x5, [sp, #32]
 0x0000000109fe4358: stp    x6, x7, [sp, #48]
 0x0000000109fe435c: stp    x8, x9, [sp, #64]
 0x0000000109fe4360: stp    x10, x11, [sp, #80]
 0x0000000109fe4364: stp    x12, x13, [sp, #96]
 0x0000000109fe4368: stp    x14, x15, [sp, #112]
 0x0000000109fe436c: stp    x16, x17, [sp, #128]
 0x0000000109fe4370: stp    x18, x19, [sp, #144]
 0x0000000109fe4374: stp    x20, x21, [sp, #160]
 0x0000000109fe4378: stp    x22, x23, [sp, #176]
 0x0000000109fe437c: stp    x24, x25, [sp, #192]
 0x0000000109fe4380: stp    x26, x27, [sp, #208]`

VladimirKempik commented:
Same thing in 17ea:
 0x00000001285942c0:  adrp  x8, 0x00000001280ed000   ;  {no_reloc}
 0x00000001285942c4:  add   x8, x8, #0x880
 0x00000001285942c8:  blr   x8
 0x00000001285942cc:  dcps1  #0xdeae
 0x00000001285942d0:  .inst  0x064e7d23 ; undefined
 0x00000001285942d4:  .inst  0x00000001 ; undefined

So i guess it's just register saving in 11, for exceptional situation.

VladimirKempik commented:
So it's
void MacroAssembler::stop(const char* msg) {
 address ip = pc();
 pusha();
 movptr(c_rarg0, (uintptr_t)(address)msg);
 movptr(c_rarg1, (uintptr_t)(address)ip);
 mov(c_rarg2, sp);
 mov(c_rarg3, CAST_FROM_FN_PTR(address, MacroAssembler::debug64));
 blr(c_rarg3);
 hlt(0);
}
where pusha is :

void MacroAssembler::pusha() {
  push(0x7fffffff, sp);
}

where push is:

// Push lots of registers in the bit set supplied.  Don't push sp.
// Return the number of words pushed
int MacroAssembler::push(unsigned int bitset, Register stack) {
  int words_pushed = 0;

  // Scan bitset to accumulate register pairs
  unsigned char regs[32];
  int count = 0;
  for (int reg = 0; reg <= 30; reg++) {
    if (1 & bitset)
      regs[count++] = reg;
    bitset >>= 1;
  }
...

@theRealAph could you please suggest, do we need to change the mask in pusha for win-arm64/mac-arm64 to not save r18 by pusha ?

theRealAph commented:
Yes, exartly.

It could be:

  `push(RegSet::range(r0, r29) R18_RESERVED_ONLY(-r18_tls), sp);`

VladimirKempik reviewed (commented):
> Yes, exartly.
> 
> It could be:
> 
> `push(RegSet::range(r0, r29) R18_RESERVED_ONLY(-r18_tls), sp);`

But it should probably be done here 
https://github.com/openjdk/jdk11u-dev/pull/301/files#diff-0f4150a9c607ccd590bf256daa800c0276144682a92bc6bdced5e8bc1bb81f3aR2548

theRealAph commented on src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp:
```diff
@@ -2544,9 +2544,17 @@ void MacroAssembler::debug64(char* msg, int64_t pc, int64_t regs[])
+  RegSet regs = RegSet::range(r0, r17) - RegSet::of(rscratch1, rscratch2);
+#ifndef R18_RESERVED
+  regs += r18_reserved;
+#endif
```
Is R18 really call clobbered? And it is Java's job to save and restore it? I would not have though it was, but I don't know the calling convention. Either way, someone should carefully review this.

theRealAph commented on src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp:
```diff
@@ -2544,9 +2544,17 @@ void MacroAssembler::debug64(char* msg, int64_t pc, int64_t regs[])
 void MacroAssembler::push_call_clobbered_registers() {
   int step = 4 * wordSize;
-  push(RegSet::range(r0, r18) - RegSet::of(rscratch1, rscratch2), sp);
+  push(call_clobbered_registers() - RegSet::of(rscratch1, rscratch2), sp);
```
rscratch1 and rscratch2 definitely are call clobbered.

theRealAph replied:
It's not exactly wrong. The readability of the code seems to have suffered in translation.

theRealAph replied:
Looking at the actual change make to development head, this is a misapplied patch. To match head, this should be:

```
RegSet MacroAssembler::call_clobbered_registers() {
  RegSet regs = RegSet::range(r0, r17) - RegSet::of(rscratch1, rscratch2);
#ifndef R18_RESERVED
  regs += r18_tls;
#endif
  return regs;
}
void MacroAssembler::push_call_clobbered_registers() {
  int step = 4 * wordSize;
  push(call_clobbered_registers(), sp);
```
Strictly speaking it should also push rscratch1 and rscratch2 - they are call clobbered - but we don't care because any macro can trash rscratch1 and rscratch2, so there's no point saving and restoring them.

rnkovacs replied:
Thanks for catching this @VladimirKempik.

rnkovacs replied:
@theRealAph you're right about `push_call_clobbered_registers()`, but I think `call_clobbered_registers()` is actually fine the way it is. The register named `r18_tls` on tip has been renamed to `r18_reserved` in this backport.

theRealAph replied:
Looks to me like the only real issue here is `warn()`. `stop()` doesn't restore the registers, so that's fine. We do want to see all the registers in a register dump. It would be better if `pusha()` were left as it is, but `popa()` did not clobber r18.

theRealAph replied:
> The register named `r18_tls` on tip has been renamed to `r18_reserved` in this backport.

Why has it been renamed `r18_reserved` ?

lewurm commented:
> [...] where pusha is :
> 
> void MacroAssembler::pusha() { push(0x7fffffff, sp); }
> 
> [...]

Nice catch!

I'm not sure if this should go as part of this PR. The same problem exists in jdk-tip and therefore should be addressed there first:
https://github.com/openjdk/jdk/blob/c05dc268acaf87236f30cf700ea3ac778e3b20e5/src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp#L1859-L1865

On jdk-tip there is only one usage left in the template interpreter generator: https://github.com/openjdk/jdk/blob/c05dc268acaf87236f30cf700ea3ac778e3b20e5/src/hotspot/cpu/aarch64/templateInterpreterGenerator_aarch64.cpp#L1400-L1404

Seems like the code path is uncommon, but we should still fix it imho.

What do you think?

VladimirKempik commented:
Since this PR is kind of stuck, can we get R18 patch out of it and push it under separate PR ? this will make macos-aarch64 port not dependent on win-aarch64 anymore

theRealAph commented:
> Since this PR is kind of stuck, can we get R18 patch out of it and push it under separate PR ? this will make macos-aarch64 port not dependent on win-aarch64 anymore

OK.

theRealAph commented:
What's going on with this patch? Seems that it includes a lot of changes for R18 that are in mainline already.

RealCLanger commented:
To unbreak things, somebody would need to review #274 as a first step.

theRealAph commented:
> To unbreak things, somebody would need to review #274 as a first step.

Done.

RealCLanger commented:
> > To unbreak things, somebody would need to review #274 as a first step.
> 
> Done.

Thanks. Maybe you could also look at #300 as I believe those two should go together. Still it's mostly about Windows.

lewurm commented:
> > Since this PR is kind of stuck, can we get R18 patch out of it and push it under separate PR ? this will make macos-aarch64 port not dependent on win-aarch64 anymore
> 
> OK.

Please let's not go there, it would delay this PR much further. I think we are close to the finish line.

Those PRs must go in the following order:
1. https://github.com/openjdk/jdk11u-dev/pull/274
2. https://github.com/openjdk/jdk11u-dev/pull/299

Then this JEP PR is ready to be merged.

rnkovacs commented:
Rebased on top of updated dependencies and removed redundant subtraction of scratch registers in `MacroAssembler::push_call_clobbered_registers()`.

I didn't do the `pusha()` part yet as I haven't seen any response to @lewurm's suggestion:

> > [...] where pusha is :
> > void MacroAssembler::pusha() { push(0x7fffffff, sp); }
> > [...]
> 
> Nice catch!
> 
> I'm not sure if this should go as part of this PR. The same problem exists in jdk-tip and therefore should be addressed there first: https://github.com/openjdk/jdk/blob/c05dc268acaf87236f30cf700ea3ac778e3b20e5/src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp#L1859-L1865
> 
> On jdk-tip there is only one usage left in the template interpreter generator: https://github.com/openjdk/jdk/blob/c05dc268acaf87236f30cf700ea3ac778e3b20e5/src/hotspot/cpu/aarch64/templateInterpreterGenerator_aarch64.cpp#L1400-L1404
> 
> Seems like the code path is uncommon, but we should still fix it imho.
> 
> What do you think?

rnkovacs commented:
> > > To unbreak things, somebody would need to review #274 as a first step.
> > 
> > 
> > Done.
> 
> Thanks. Maybe you could also look at #300 as I believe those two should go together. Still it's mostly about Windows.

Thank you @theRealAph!

Actually, the connection between #274 and #300 is administrative in nature (originally, refactorings of 3 different areas were reviewed under the same JBS bug), there's no real dependency. It doesn't have to delay this stack of patches.

rnkovacs commented:
@RealCLanger responding to [your comment](https://github.com/openjdk/jdk11u-dev/pull/222#issuecomment-933834090) under #222, the remaining PRs:

These 3 need to go in in this order:
1. #274 (**accepted**, waiting for JBS `fix-yes` tag)
2. #299 (informally accepted by original author, waiting for formal review)
3. #301 (this one)

These two depend on #301:
- #304 (**accepted**)
- #307 (waiting for review)

These two could go in independently:
- #300 (**accepted** but debated)
- #280 (informally accepted by original author, waiting for formal review)

VladimirKempik commented:
> > [...] where pusha is :
> > void MacroAssembler::pusha() { push(0x7fffffff, sp); }
> > [...]
> 
> Nice catch!
> 
> I'm not sure if this should go as part of this PR. The same problem exists in jdk-tip and therefore should be addressed there first: https://github.com/openjdk/jdk/blob/c05dc268acaf87236f30cf700ea3ac778e3b20e5/src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp#L1859-L1865
> 
> On jdk-tip there is only one usage left in the template interpreter generator: https://github.com/openjdk/jdk/blob/c05dc268acaf87236f30cf700ea3ac778e3b20e5/src/hotspot/cpu/aarch64/templateInterpreterGenerator_aarch64.cpp#L1400-L1404
> 
> Seems like the code path is uncommon, but we should still fix it imho.
> 
> What do you think?

I haven't spotted the usage of r18 ( saving r18 on stack) in jdk17 with hsdis, hence I was thinking it's not affected and the fix is needed only in jdk11

rnkovacs commented:
Rebase after the merge of #274.

theRealAph commented:
> > _Mailing list message from [Vladimir Kempik](mailto:vkempik@azul.com) on [jdk-updates-dev](mailto:jdk-updates-dev@mail.openjdk.java.net):_
> > On one side - we should be as close to original commit in jdk/jdk as possible on another side - r18_tls sounds windows-aarch64 only, r18_reserved sounds more general ( works for both macos/win).
> 
> Thanks @VladimirKempik, that's the reason why I assumed `r18_reserved` was the newer one. Which one do you prefer @theRealAph?

`r18_tls`. It's not very important, bur don't think the name is inaccurate: as far as I know both platforms use r18 for thread-local storage. It' could just have been `r18_do_not_use`.

VladimirKempik commented:
> > > _Mailing list message from [Vladimir Kempik](mailto:vkempik@azul.com) on [jdk-updates-dev](mailto:jdk-updates-dev@mail.openjdk.java.net):_
> > > On one side - we should be as close to original commit in jdk/jdk as possible on another side - r18_tls sounds windows-aarch64 only, r18_reserved sounds more general ( works for both macos/win).
> > 
> > 
> > Thanks @VladimirKempik, that's the reason why I assumed `r18_reserved` was the newer one. Which one do you prefer @theRealAph?
> 
> `r18_tls`. It's not very important, but Idon't think the name is inaccurate: as far as I know both platforms use r18 for thread-local storage. It' could just have been `r18_do_not_use`.

Oh well, AFAIK nobody outside Apple knows what it's used for. It's even possible to use R18 in your own app on macos11/12 ( except macos12b6 - https://youtrack.jetbrains.com/issue/JBR-3715 )
By mistake we left r18 available for C2 on zulu11/13/15 ( also JetBrains Runtime 11) and it was working fine on all macos versions except 12beta6.

rnkovacs commented:
Rebased on top of #567

theRealAph commented:
I don't understand the build machinery enough to comment. Maybe someone who does should have a look.

theRealAph reviewed (commented):
Looks good. We'll need another reviewer.

theRealAph commented on make/autoconf/toolchain.m4:
```diff
@@ -918,14 +918,18 @@ AC_DEFUN_ONCE([TOOLCHAIN_SETUP_BUILD_COMPILERS],
           # Corresponds to --with-toolchain-path
-          BASIC_EVAL_DEVKIT_VARIABLE([BUILD_DEVKIT_TOOLCHAIN_PATH])
+          BASIC_EVAL_BUILD_DEVKIT_VARIABLE([BUILD_DEVKIT_TOOLCHAIN_PATH])
           # Corresponds to --with-sysroot
```
OK.

lewurm commented:
> I don't understand the build machinery enough to comment. Maybe someone who does should have a look.

@magicus could you help out with this?

lewurm commented on src/hotspot/cpu/aarch64/aarch64.ad:
```diff
@@ -5433,7 +5433,7 @@ pipeline %{
   // ARM instructions are of fixed length
   fixed_size_instructions;        // Fixed size instructions TODO does
-  max_instructions_per_bundle = 2;   // A53 = 2, A57 = 4
+  max_instructions_per_bundle = 4;   // A53 = 2, A57 = 4
```
I don't recall the details, but it was part of the JEP in tip: https://github.com/openjdk/jdk/pull/212/files#diff-018aa61d1a7aafcf70a535fcd40a318a4bd6511fd40ac39ce4be90cc52216749R6297

@mo-beck do you remember the motivation behind this change?

theRealAph replied:
Please take it out, then. Whatever the reason might have been, we don't need it here.

rnkovacs replied:
Rolled back in https://github.com/openjdk/jdk11u-dev/pull/301/commits/c0073dad2a6bb1c612d63f2572d67f9254460438.

magicus reviewed (commented):
Apart from the (not so minor!) issue that the entire windows/aarch64 cannot be built unless the user provides a fixpath.exe binary in the repository root, the the build changes look okay, I guess. I can only wonder a bit how this is supposed to be solved. Even if a local developer can supply a custom-made fixpath.exe, this will prevent e.g. the automatic testing of windows/aarch64 on github actions (if you have those enabled for 11u).

Most of the changes would only affect windows/aarch64, and since that is a new platform, the risk to existing 11u platforms is zero for those changes. However, there are also some changes that will affect all platforms, or a subset of them (like Windows, or aarch64 in general). I *think* that these changes are okay, but I can't guarantee; I do not readily remember the details of the jdk 11 build code base vs the mainline.

If one wanted to be more cautious, a safer route is probable to leave the current platforms unaffected, and duplicate the code with the new implementation just for windows/aarch64.

But then again, I don't understand the JDK updates criteria anymore. I thought updates were for critical bug fixes only, not adding entirely new platforms.

theRealAph commented on make/autoconf/basics_windows.m4:
```diff
@@ -396,22 +430,7 @@ AC_DEFUN_ONCE([BASIC_COMPILE_FIXPATH],
-      cat $FIXPATH_DIR/fixpath1.log
-      AC_MSG_ERROR([Could not create $FIXPATH_BIN])
-    fi
-    AC_MSG_RESULT([yes])
```
> So how are you supposed to get the fixpath.exe on windows/aarch64?
> 
> On the surface, this backport seem to have the same problem as your original win/aarch64 patch for mainline, that it is not generally possible to build it without a lot of external hacks. I can only point this out; I'll have to leave it to the jdk11u maintainers to accept or drop the patch given this knowledge.

Auto-testing is always going to be a pain as long as the port isn't bootstrappable because there aren't native tools, right?

lewurm replied:
> So how are you supposed to get the fixpath.exe on windows/aarch64?

If you do not want to download an arbitrary `.exe` file as described mailing list post linked above (which I totally get), you can just do a build for Windows/x86 and copy over the `fixpath.exe` that was generated by that.

I understand having hacks around the build system issues is less than ideal. Ideally we would backport your WINENV patch to 11, but that would be much more invasive in terms of risk managment. The proposed solution in this PR does at least barely touch other platforms, so the risk for breaking something is rather low imho.

theRealAph commented:
> But then again, I don't understand the JDK updates criteria anymore. I thought updates were for critical bug fixes only, not adding entirely new platforms.

I thing it's been a pretty long time since we were *that* strict. Longer than I can remember, anyway.
The rule of thumb is that if we want older JDK releases to remain relevant, they have to be able to adapt to new platforms. So, we change JDK code for OS updates, compiler and library changes, and so on. Adding Windows/AArch64 is a pretty extreme case of that, but it's within the rule, but only just. After all, we already had Windows and we already had AArch64.

rnkovacs commented:
Removed changes pointed out as unnecessary in a new commit: `max_instructions_per_bundle` in `aarch64.ad` and the `machine` flag removals in `flags-ldflags.m4`.

(I did not change the backport commit, sorry about the force-push - I just rebased to master before realizing that #567 is not integrated yet, so I had to walk back on that.)

theRealAph commented on make/autoconf/flags-ldflags.m4:
```diff
@@ -186,15 +186,14 @@ AC_DEFUN([FLAGS_SETUP_LDFLAGS_CPU_DEP],
-    if test "x${OPENJDK_$1_CPU}" = "xx86"; then
-      $1_CPU_LDFLAGS="-safeseh"
-      # NOTE: Old build added -machine. Probably not needed.
-      $1_CPU_LDFLAGS_JVM_ONLY="-machine:I386"
```
You're right, it is not prudent to do this.

rnkovacs replied:
Switched back to the old version, just adding an aarch64-specific branch in https://github.com/openjdk/jdk11u-dev/pull/301/commits/c0073dad2a6bb1c612d63f2572d67f9254460438.

theRealAph commented on src/hotspot/os_cpu/windows_aarch64/bytes_windows_aarch64.inline.hpp:
```diff
@@ -0,0 +1,46 @@
+  return _byteswap_uint64(x);
+}
+
+#pragma warning(default: 4035) // Enable warning 4035: no return value
```
What on Earth is this line doing here?

lewurm replied:
Probably a copy-pasta leftover from the x86 bit (where it actually makes sense): https://github.com/openjdk/jdk11u-dev/blob/1ae15632c7005eff2267aa6cdbacbc038c4b43c5/src/hotspot/os_cpu/windows_x86/bytes_windows_x86.inline.hpp#L85

rnkovacs replied:
Removed. I'll take a note to remove it on tip too.

rnkovacs commented:
Rebase to master again.

theRealAph commented on src/hotspot/cpu/aarch64/c1_FpuStackSim_aarch64.cpp:
```diff
@@ -28,3 +28,4 @@
 //--------------------------------------------------------
 
 // No FPU stack on AARCH64
+#include "precompiled.hpp"
```
Why does Windows need this change?

rnkovacs replied:
It seems that MSVC gives an [error](https://docs.microsoft.com/en-us/cpp/error-messages/compiler-errors-1/fatal-error-c1010?view=msvc-170) if a PCH specified on the command line is not included in the file. Since the use of precompiled headers is enabled by default, this compilation command gets the same PCH related arguments as the others. An alternative solution I found to be working is to add this file to the PCH exclude list here: [make/hotspot/lib/JvmOverrideFiles.gmk#L191](https://github.com/openjdk/jdk11u-dev/blob/6ec651fdc1c5ea0fe22731077446de1dd39cb24d/make/hotspot/lib/JvmOverrideFiles.gmk#L191), and then the include can be removed. Which one do you prefer?

theRealAph replied:
This change is OK, then. Let it stay.

adinn replied:
It would help avoid problems later if a comment was added to say that this include is needed to keep MSVC happy. Otherwise someone might delete it.

theRealAph reviewed (approved):
Thanks. That looks pretty clean to me now.
I think we need one more reviewer, and we're good.

rnkovacs commented:
> Thanks. That looks pretty clean to me now. I think we need one more reviewer, and we're good.

Thanks so much for the review, @theRealAph.
Could you please suggest someone as the other reviewer?

theRealAph commented:
> Thanks so much for the review, @theRealAph. Could you please suggest someone as the other reviewer?

@adinn, I guess. I'm happy enough that Microsoft can check the Windows parts.

rnkovacs commented:
> > Thanks so much for the review, @theRealAph. Could you please suggest someone as the other reviewer?
> 
> @adinn, I guess. I'm happy enough that Microsoft can check the Windows parts.

Thanks. @adinn, could you please take a look?

Also, reaching back to the original authors one last time (thanks @lewurm for sticking around through the whole process) - @luhenry and @mo-beck, does the backport look good to you?

adinn commented:
@rnkovacs I have started reviewing this PR.

rnkovacs commented:
Good idea, I added the comment. Thanks for the review, @adinn.

rnkovacs commented:
Some fresh test results for the two Arm64 platforms:

**Linux-Arm64** fastdebug before the patch:

  | Total | Passed | Failed | Error
-- | -- | -- | -- | --
jdk tier1 | 1845 | 1845 | 0 | 0
hotspot tier1 | 1514 | 1514 | 0 | 0
langtools all | 3926 | 3926 | 0 | 0

**Linux-Arm64** fastdebug after the patch:

  | Total | Passed | Failed | Error
-- | -- | -- | -- | --
jdk tier1 | 1845 | 1845 | 0 | 0
hotspot tier1 | 1514 | 1514 | 0 | 0
langtools all | 3926 | 3926 | 0 | 0

**Windows-Arm64** fastdebug:

  | Total | Passed | Failed | Error
-- | -- | -- | -- | --
jdk tier1 | 1865 | 1864 | 0 | 0
hotspot tier1 | 1246 | 1238 | 8 | 0
langtools all | 3925 | 3925 | 0 | 0

[Here](https://gist.github.com/rnkovacs/edfa07b7b86ae720c26e788db230f872)'s a list of the failing hotspot tests.

We've been battling issues with both Cygwin and MSVC. We've reported a [bad arm64 codegen bug](https://developercommunity.visualstudio.com/t/arm64-incorrect-code-gen-when-compiler-decides-to/1541140) that causes assert failures in a bunch of tests, and is still present in VS 2022 (closed as fixed incorrectly - we're contacting the engineers directly). We got our results by using a [workaround](https://gist.github.com/rnkovacs/8b67a8513d19b43343185dc5f86fe3fe) that seems to fix the issue.

rnkovacs commented:
I've added the `fix-request` tag to the JBS issue.

theRealAph commented:
> [Here](https://gist.github.com/rnkovacs/edfa07b7b86ae720c26e788db230f872)'s a list of the failing hotspot tests.
> 
> We've been battling issues with both Cygwin and MSVC. We've reported a [bad arm64 codegen bug](https://developercommunity.visualstudio.com/t/arm64-incorrect-code-gen-when-compiler-decides-to/1541140) that causes assert failures in a bunch of tests, and is still present in VS 2022 (closed as fixed incorrectly - we're contacting the engineers directly). We got our results by using a [workaround](https://gist.github.com/rnkovacs/8b67a8513d19b43343185dc5f86fe3fe) that seems to fix the issue.

I don't see any reason why we'd reject your workaround, so please add it to this PR.

I'll approve this _pro tem_  without the Serviceability fixes, but I don't want them to be neglected. Please let us know about your plans to get Serviceability done.

adinn commented:
What's up with the first line of the Windows-Arm64 fastdebug test results? The number of jdk tier1 tests that passed is one short of the total but no failures or errors are listed. Any explanation?

rnkovacs commented:
@adinn Sorry for that mistake. There is one failing test there, `java/lang/ProcessBuilder/InheritIO/InheritIO.sh`, but it's an environment issue. The test expects to see `exit value: 0` in stderr, but instead it gets `Cygwin WARNING: (...) exit value: 0`. Apparently the warning could be prevented by using either a newer Cygwin or an older Windows. We've updated our Cygwin packages to 3.2.0 as the latest stable version, and it does solve the issue e.g. paired with an x64 Windows Server 2016, but our Arm64 machines all have OSes from this year, and we'd like to avoid downgrading (we had problems with previous builds).

rnkovacs commented:
> I don't see any reason why we'd reject your workaround, so please add it to this PR.
> 
> I'll approve this _pro tem_ without the Serviceability fixes, but I don't want them to be neglected. Please let us know about your plans to get Serviceability done.

Thank you - I added it in a new commit.

I got reassured that fixing these failures is a top priority for us, especially since we see most of them with our JDK 17 builds as well. We can start investigating right away, but we'd like to merge this first to make sure it's not delayed any longer.

adinn commented:
@rnkovacs Ok, no problem with that test failure then. Good to merge then.

theRealAph commented:
> @rnkovacs Ok, no problem with that test failure then. Good to merge then.

We're getting very close to freeze date for 11u release. Is this a good time?

adinn commented:
> We're getting very close to freeze date for 11u release. Is this a good time?

Sorry, I just meant good as far as I am concerned. Not trying to usurp the maintainer role.

theRealAph commented:
> > We're getting very close to freeze date for 11u release. Is this a good time?
> 
> Sorry, I just meant good as far as I am concerned. Not trying to usurp the maintainer role.

Yea, I know. Was just asking.

adinn commented:
> Yea, I know. Was just asking.

Well you're the maintainer ;-)

As a reviewer, I think the patch does an excellent job of minimizing shared/alternative port changes and ensuring the the ones it does make are safe. But it does need to make such changes. So, albeit it's a good, reliable bit of work but there is still a small risk of problems.

It's up to the (other?) maintainers to make the call here.

GoeLin commented:
Hi,
I would vote for approving the change. Why?
- It is well reviewed.
- The shared changes are mostly procedural changes resolving dependencies but not affecting the core functionality (interpreter, jit, runtime…). Obviously, the windows x86/x64 and linux aarch64 ports are affected more by this change than the other ports, as there are a row of files touched because they are now shared with the windows aarch64 port.  But this also limits the risk to a subset of the ports.
- There are still four weeks rampdown, and basically it is the idea of rampdown to test the changes approved and pushed before.
- The change is live upstream for quite a while now, so issues should only result from adaptions to 11, there should not be an unexpected high ratio of uncaught basic issues in the port.
- Availability of a new port is valuable.

theRealAph commented:
> Hi, I would vote for approving the change. Why?
> 
>     * It is well reviewed.
> 
>     * The shared changes are mostly procedural changes resolving dependencies but not affecting the core functionality (interpreter, jit, runtime…). Obviously, the windows x86/x64 and linux aarch64 ports are affected more by this change than the other ports, as there are a row of files touched because they are now shared with the windows aarch64 port.  But this also limits the risk to a subset of the ports.
> 
>     * There are still four weeks rampdown, and basically it is the idea of rampdown to test the changes approved and pushed before.
> 
>     * The change is live upstream for quite a while now, so issues should only result from adaptions to 11, there should not be an unexpected high ratio of uncaught basic issues in the port.
> 
>     * Availability of a new port is valuable.

That sounds right, thanks.

rnkovacs commented:
Thank you all so much for taking the time to review the changes, we really appreciate it.

rnkovacs commented:
/integrate

mo-beck commented:
Thank you all for your thoroughness, support, guidance, and perseverance throughout.

VladimirKempik commented:
Could you /reviewer credit me too? Thanks

rnkovacs commented:
/reviewer credit @VladimirKempik 
/reviewer credit @magicus

rnkovacs commented:
> Could you /reviewer credit me too? Thanks

Sorry I thought you are. It seems like it picks up people who actually approve the PR.

theRealAph commented:
> > Could you /reviewer credit me too? Thanks
> 
> Sorry I thought you are. It seems like it picks up people who actually approve the PR.

I'm looking at the OpenJDK census, and vkempik seems to be a committer, not a reviewer. It would be quite wrong for us to record a non-reviewer as a reviewer.

rnkovacs commented:
> I'm looking at the OpenJDK census, and vkempik seems to be a committer, not a reviewer. It would be quite wrong for us to record a non-reviewer as a reviewer.

Should I remove @lewurm and @mo-beck from the reviewers then? The tooling itself seems to be confused about GH reviews from people with and without Reviewer status, as they got added automatically.

rnkovacs commented:
/reviewer remove burban
/reviewer remove mbeckwit

rnkovacs commented:
I cannot do any meaningful modifications to the reviewer list.

Could someone please sponsor the change?

RealCLanger commented:
> > > Could you /reviewer credit me too? Thanks
> > 
> > 
> > Sorry I thought you are. It seems like it picks up people who actually approve the PR.
> 
> I'm looking at the OpenJDK census, and vkempik seems to be a committer, not a reviewer. It would be quite wrong for us to record a non-reviewer as a reviewer.

No, everybody who reviews a PR (marks the PR as reviewed) and has a census entry will be credited as a reviewer by Skara. But only if somebody is a project Reviewer - by census role - this will be an authoritative review, allowing the author to merge a PR. And I also think that's right, anybody who spent time and effort to review a particular change should also be credited, no matter of project role.

RealCLanger commented:
> I cannot do any meaningful modifications to the reviewer list.
> 
> Could someone please sponsor the change?

OK, I hope @VladimirKempik and @magicus will be credited correctly, but they should. The preliminary commit message in the commit above is hopefully misleading. So:
/integrate

RealCLanger commented:
> OK, I hope @VladimirKempik and @magicus will be credited correctly, but they should. The preliminary commit message in the commit above is hopefully misleading.

Hm, my hope was wrong, I'm sorry, @VladimirKempik. I now think you would have had to approve the PR in GitHub once again...

rnkovacs commented:
Thanks so much, @RealCLanger!

brianjstafford commented:
Thank you to all the contributors and reviewers who have helped with this PR. It's very much appreciated!

theRealAph commented:
> No, everybody who reviews a PR (marks the PR as reviewed) and has a census entry will be credited as a reviewer by Skara. But only if somebody is a project Reviewer - by census role - this will be an authoritative review, allowing the author to merge a PR. And I also think that's right, anybody who spent time and effort to review a particular change should also be credited, no matter of project role.

Yes, I get that, but it's a trap for all contributors: if you go to "Reviewers" at the top of the page you just see a list. I've got nothing against credit where it's due, but this workflow looks rather error prone.

mo-beck commented:
I agree with @RealCLanger. The original work was done by 3 contributors, so I think it is appreciable (and should be encouraged) if any of the original contributors spent their time reviewing the backport, and hence they should be listed as reviewers (despite their project status). Similarly, @VladimirKempik and @magicus spent their time and effort helping out with the original work and the backport as well. For some reason, I don't see them both in the reviewer for the backport and I don't see any of their code comments when I click on their `review comments` icon on the right top. So maybe @VladimirKempik and @magicus would need to manually mark this backport as reviewed to get the credit. Thanks for having this discussion.

> > > > Could you /reviewer credit me too? Thanks
> > > 
> > > 
> > > Sorry I thought you are. It seems like it picks up people who actually approve the PR.
> > 
> > 
> > I'm looking at the OpenJDK census, and vkempik seems to be a committer, not a reviewer. It would be quite wrong for us to record a non-reviewer as a reviewer.
> 
> No, everybody who reviews a PR (marks the PR as reviewed) and has a census entry will be credited as a reviewer by Skara. But only if somebody is a project Reviewer - by census role - this will be an authoritative review, allowing the author to merge a PR. And I also think that's right, anybody who spent time and effort to review a particular change should also be credited, no matter of project role.

VladimirKempik commented:
probably it's github issue, the link that says %Username% left review comments is not working anymore for me and magicus
an example https://github.com/openjdk/jdk11u-dev/pull/301/files/f7f50ab7bdaa0c6f893d0e0874de25a83b7d2edd 
So I can't mark those comments as fixed and "approve" button is nowhere to be found.

theRealAph commented:
> I agree with @RealCLanger. The original work was done by 3 contributors, so I think it is appreciable (and should be encouraged) if any of the original contributors spent their time reviewing the backport, and hence they should be listed as reviewers (despite their project status).

Sure, it's a joint contribution, and all authors should be credited.

> Similarly, @VladimirKempik and @magicus spent their time and effort helping out with the original work and the backport as well. For some reason, I don't see them both in the reviewer for the backport and I don't see any of their code comments when I click on their `review comments` icon in the top right area. So maybe @VladimirKempik and @magicus would need to manually mark this backport as reviewed to get the credit. Thanks for having this discussion.

Fair enough, but that's about acquiring brownie points, whereas I'm talking about processes. 

The purpose of a review tick is to mark a change as reviewed, and we need Reviewers to do that. However, the list of "Reviewers" at the top right of this page doesn't indicate which ones are Reviewers. Thankfully, Skara does provide the information we need (in the Reviewers section at the top of the discussion) so we have what we need.
